### PR TITLE
Process inline code generated by another template if annotation's suffix is one of the templates in force-parse argument

### PR DESCRIPTION
**Pull request by ranvirp**
There is requirement that inline code generated from a different template can be processed. force-parse arguments provide this ability for generated code in a file. In this pull request, annotationranges is given forceparse argument and if the inline annotation ends with one of the templates then that inline code is not removed.

### CODE REVIEW

File: SourceryFramework/Sources/Parsing/FileParser.swift [hunk 1]
```diff
@@ -56,6 +56,7 @@ public final class FileParser {
     fileprivate var annotations: AnnotationsParser!
     fileprivate var inlineRanges: [String: NSRange]!
     fileprivate var inlineIndentations: [String: String]!
+    fileprivate var forceParse:[String] = ["AutoCoding"]
```
Comment (krzysztofzablocki) [inline SourceryFramework/Sources/Parsing/FileParser.swift]: This and initializer default value should be removed IMO

File: SourceryFramework/Sources/Parsing/Utils/InlineParser.swift [hunk 2]
```diff
@@ -60,7 +60,22 @@ public enum TemplateAnnotationsParser {
...
+                
+                
+                if name.hasSuffix("." + ignoreTemplate) {
+                    rangeToBeRemoved = false
+                    break
+                }
+            }
+            
+            if rangeToBeRemoved || name.hasSuffix("." + "AutoCoding") {
+             //   print("removed named range \(name) and \(forceParse)")
```
Comment (krzysztofzablocki) [inline SourceryFramework/Sources/Parsing/Utils/InlineParser.swift]: lets remove this commented out code

**Comment by krzysztofzablocki** [comment]
@ranvirp can you update the changes I requested?

File: SourceryFramework/Sources/Parsing/Utils/InlineParser.swift (hunk 2 above)
Comment (ilyapuchka) [reply SourceryFramework/Sources/Parsing/Utils/InlineParser.swift]: I'm also not sure why AutoCoding is explicitly used here

File: SourceryTests/Parsing/TemplateAnnotationsParserSpec.swift [hunk 3]
```diff
@@ -46,7 +46,7 @@ class TemplateAnnotationsParserSpec: QuickSpec {
                         "    var something: Int\n" +
                         "    // sourcery:end\n"
 
-                let result = TemplateAnnotationsParser.parseAnnotations("inline", contents: source)
+                let result = TemplateAnnotationsParser.parseAnnotations("inline", contents: source, aggregate: false, forceParse: [])
```
Comment (ilyapuchka) [inline SourceryTests/Parsing/TemplateAnnotationsParserSpec.swift]: We should have some tests that validate expectations when this attribute is set to something.

**Comment by ilyapuchka** [comment]
tbh I couldn't get how this would be actually used by the users, it seems like implementation was specifically made to handle parsing code generated for AutoCodable.

File: SourceryFramework/Sources/Parsing/Utils/InlineParser.swift (hunk 2 above)
Comment (ranvirp) [reply SourceryFramework/Sources/Parsing/Utils/InlineParser.swift:75]: How to add changelog?

File: SourceryFramework/Sources/Parsing/FileParser.swift (hunk 1 above)
Comment (ranvirp) [reply SourceryFramework/Sources/Parsing/FileParser.swift:59]: Without a default value, all tests would need to be changed.

File: SourceryFramework/Sources/Parsing/Utils/InlineParser.swift (hunk 2 above)
Comment (ranvirp) [reply SourceryFramework/Sources/Parsing/Utils/InlineParser.swift]: > I'm also not sure why AutoCoding is explicitly used here
Purpose of using a default value is to allow all tests to pass.

File: SourceryFramework/Sources/Parsing/FileParser.swift (hunk 1 above)
Comment (ranvirp) [reply SourceryFramework/Sources/Parsing/FileParser.swift]: Default Value removed

File: SourceryTests/Parsing/TemplateAnnotationsParserSpec.swift (hunk 3 above)
Comment (ranvirp) [reply SourceryTests/Parsing/TemplateAnnotationsParserSpec.swift:49]: > We should have some tests that validate expectations when this attribute is set to something.

Test added.

**Comment by krzysztofzablocki** [comment]
@ranvirp can you comment on @ilyapuchka question?

**Comment by ranvirp** [comment]
> tbh I couldn't get how this would be actually used by the users, it seems like implementation was specifically made to handle parsing code generated for AutoCodable.

Now Autocoding default removed. And test also added. This feature is super useful when inline code generated by one Sourcery template has to be processed by subsequent templates. Without it, the code inside inline annotations are deleted. For example, It is useful when one template generates enums and another template generates tests based on those enum.

File: README.md [hunk 4]
```diff
@@ -107,7 +107,7 @@ $ ./sourcery --sources <sources path> --templates <templates path> --output <out
 
 - `--sources` - Path to a source swift files or directories. You can provide multiple paths using multiple `--sources` option.
 - `--templates` - Path to templates. File or Directory. You can provide multiple paths using multiple `--templates` options.
-- `--force-parse` - File extensions of Sourcery generated file you want to parse. You can provide multiple extension using multiple `--force-parse` options. (i.e. `file.toparse.swift` will be parsed even if generated by Sourcery if `--force-parse toparse`). Useful when trying to implement a multiple phases generation.
+- `--force-parse` - File extensions of Sourcery generated file you want to parse. You can provide multiple extension using multiple `--force-parse` options. (i.e. `file.toparse.swift` will be parsed even if generated by Sourcery if `--force-parse toparse`). Useful when trying to implement a multiple phases generation. Any inline code with annotation ending with "." and this option would also be parsed(i.e. inline code beginning with // sourcery:inline:auto:Command.MirrorEnum would be parsed even if generated by Sourcery if `--force-parse MirrorEnum`).
```
Comment (ilyapuchka) [inline README.md:110]: Can we rephrase it? I don't get what "ending with ." means from this description and tbh I find original description quite confusing as well.

```suggestion
- `--force-parse` - normally Sourcery does not process files or code within inline annotations generated by Sourcery itself. To force parsing of such files or code within inline annotations pass the extension (without `.swift`) of the file or the name of the annotation to this option. I.e. to process file `*.generated.swift` you can use `--force-parse generated`, to process code within `sourcery:inline:auto:Type.AutoCodable` annotation you can use `--force-parse AutoCodable`. You can provide this option multiple times. Useful when trying to implement a multiple phases generation.
```

@krzysztofzablocki does this describe it correctly? I'm not sure I understand how ti works on files.
I'm also not sure using one option for both files and annotations is a good idea.

File: SourceryFramework/Sources/Parsing/FileParser.swift [hunk 5]
```diff
@@ -56,6 +56,7 @@ public final class FileParser {
     fileprivate var annotations: AnnotationsParser!
     fileprivate var inlineRanges: [String: NSRange]!
     fileprivate var inlineIndentations: [String: String]!
+    fileprivate var forceParse:[String] = []
```
Comment (ilyapuchka) [inline SourceryFramework/Sources/Parsing/FileParser.swift:59]:
```suggestion
    fileprivate var forceParse: [String] = []
```

File: SourceryFramework/Sources/Parsing/FileParser.swift [hunk 6]
```diff
@@ -68,11 +69,12 @@ public final class FileParser {
     ///   - contents: Contents to parse.
     ///   - path: Path to file.
     /// - Throws: parsing errors.
-    public init(contents: String, path: Path? = nil, module: String? = nil) throws {
+    public init(contents: String, path: Path? = nil, module: String? = nil,forceParse:[String]=[]) throws {
```
Comment (ilyapuchka) [inline SourceryFramework/Sources/Parsing/FileParser.swift:72]:
```suggestion
    public init(contents: String, path: Path? = nil, module: String? = nil, forceParse: [String] = []) throws {
```

File: SourceryFramework/Sources/Parsing/FileParser.swift [hunk 7]
```diff
@@ -83,7 +85,7 @@ public final class FileParser {
             return contents
         }
 
-        let inline = TemplateAnnotationsParser.parseAnnotations("inline", contents: initialContents)
+        let inline = TemplateAnnotationsParser.parseAnnotations("inline", contents: initialContents,aggregate: false,forceParse: self.forceParse)
```
Comment (ilyapuchka) [inline SourceryFramework/Sources/Parsing/FileParser.swift:88]:
```suggestion
        let inline = TemplateAnnotationsParser.parseAnnotations("inline", contents: initialContents, aggregate: false, forceParse: self.forceParse)
```

File: SourceryFramework/Sources/Parsing/Utils/InlineParser.swift [hunk 8]
```diff
@@ -18,8 +18,8 @@ public enum TemplateAnnotationsParser {
         return regex
     }
 
-    public static func parseAnnotations(_ annotation: String, contents: String, aggregate: Bool = false) -> (contents: String, annotatedRanges: AnnotatedRanges) {
-        let (annotatedRanges, rangesToReplace) = annotationRanges(annotation, contents: contents, aggregate: aggregate)
+    public static func parseAnnotations(_ annotation: String, contents: String, aggregate: Bool = false,forceParse:[String]) -> (contents: String, annotatedRanges: AnnotatedRanges) {
```
Comment (ilyapuchka) [inline SourceryFramework/Sources/Parsing/Utils/InlineParser.swift:21]:
```suggestion
    public static func parseAnnotations(_ annotation: String, contents: String, aggregate: Bool = false, forceParse: [String]) -> (contents: String, annotatedRanges: AnnotatedRanges) {
```

File: SourceryFramework/Sources/Parsing/Utils/InlineParser.swift [hunk 9]
```diff
@@ -18,8 +18,8 @@ public enum TemplateAnnotationsParser {
...
+        let (annotatedRanges, rangesToReplace) = annotationRanges(annotation, contents: contents, aggregate: aggregate,forceParse: forceParse)
```
Comment (ilyapuchka) [inline SourceryFramework/Sources/Parsing/Utils/InlineParser.swift:22]:
```suggestion
        let (annotatedRanges, rangesToReplace) = annotationRanges(annotation, contents: contents, aggregate: aggregate, forceParse: forceParse)
```

File: SourceryFramework/Sources/Parsing/Utils/InlineParser.swift [hunk 10]
```diff
@@ -30,7 +30,7 @@ public enum TemplateAnnotationsParser {
         return (bridged as String, annotatedRanges)
     }
 
-    public static func annotationRanges(_ annotation: String, contents: String, aggregate: Bool = false) -> (annotatedRanges: AnnotatedRanges, rangesToReplace: Set<NSRange>) {
+    public static func annotationRanges(_ annotation: String, contents: String, aggregate: Bool = false,forceParse:[String]) -> (annotatedRanges: AnnotatedRanges, rangesToReplace: Set<NSRange>) {
```
Comment (ilyapuchka) [inline SourceryFramework/Sources/Parsing/Utils/InlineParser.swift:33]:
```suggestion
    public static func annotationRanges(_ annotation: String, contents: String, aggregate: Bool = false, forceParse: [String]) -> (annotatedRanges: AnnotatedRanges, rangesToReplace: Set<NSRange>) {
```

File: SourceryFramework/Sources/Parsing/Utils/InlineParser.swift [hunk 11]
```diff
@@ -60,7 +60,21 @@ public enum TemplateAnnotationsParser {
             } else {
                 annotatedRanges[name] = [(range: range, indentation: indentation)]
             }
-            rangesToReplace.insert(range)
+            var rangeToBeRemoved = true
+            
+            for ignoreTemplate in forceParse {
+                
+
```
Comment (ilyapuchka) [inline SourceryFramework/Sources/Parsing/Utils/InlineParser.swift:67]:
```suggestion
```

File: SourceryFramework/Sources/Parsing/Utils/InlineParser.swift [hunk 12]
```diff
@@ -60,7 +60,21 @@ public enum TemplateAnnotationsParser {
...
+                
+                if name.hasSuffix("." + ignoreTemplate) {
+                    rangeToBeRemoved = false
+                    break
+                }
+            }
+            
+            if rangeToBeRemoved {
+                rangesToReplace.insert(range)
+
```
Comment (ilyapuchka) [inline SourceryFramework/Sources/Parsing/Utils/InlineParser.swift:76]:
```suggestion
```

File: SourceryFramework/Sources/Parsing/Utils/InlineParser.swift [hunk 13]
```diff
@@ -60,7 +60,21 @@ public enum TemplateAnnotationsParser {
...
+                if name.hasSuffix("." + ignoreTemplate) {
+                    rangeToBeRemoved = false
+                    break
+                }
+            }
+            
+            if rangeToBeRemoved {
+                rangesToReplace.insert(range)
+                
+            }
```
Comment (ilyapuchka) [inline SourceryFramework/Sources/Parsing/Utils/InlineParser.swift:77]:
```suggestion
            rangeToBeRemoved = !forceParse.contains(where: { name.hasSuffix("." + $0) })
            if rangeToBeRemoved {
                rangesToReplace.insert(range)
            }
```

File: SourceryTests/Parsing/TemplatesAnnotationParser+ForceParseInlineCodeSpec.swift [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+class TemplatesAnnotationParserPassInlineCodeSpec: QuickSpec {
+    override func spec() {
+        describe("InlineParser") {
+            context("without indentation") {
+                let source =
+                    "// sourcery:inline:Type.AutoCoding\n" +
+                        "var something: Int\n" +
+                "// sourcery:end\n"
```
Comment (ilyapuchka) [inline SourceryTests/Parsing/TemplatesAnnotationParser+ForceParseInlineCodeSpec.swift:28]:
```suggestion
                let source = """
                    // sourcery:inline:Type.AutoCoding
                        var something: Int
                    // sourcery:end
                    """
```

File: SourceryTests/Parsing/TemplatesAnnotationParser+ForceParseInlineCodeSpec.swift [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+                    let annotatedRanges = result.annotatedRanges["Type.AutoCoding"]
+                    expect(annotatedRanges?.map { $0.range }).to(equal([NSRange(location: 35, length: 19)]))
+                    expect(annotatedRanges?.map { $0.indentation }).to(equal([""]))
+                }
+
+                it("does not remove content between the markup when force parse parameter is set to template name") {
+                    expect(result.contents).to(equal(
+                        "// sourcery:inline:Type.AutoCoding\n" +
+                            "var something: Int\n" +
+                        "// sourcery:end\n"
```
Comment (ilyapuchka) [inline SourceryTests/Parsing/TemplatesAnnotationParser+ForceParseInlineCodeSpec.swift:42]:
```suggestion
                    expect(result.contents).to(equal("""
                        // sourcery:inline:Type.AutoCoding
                            var something: Int
                        // sourcery:end
                        """
```

File: SourceryTests/Parsing/TemplatesAnnotationParser+ForceParseInlineCodeSpec.swift [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+                    ))
+                }
+            }
+
+            context("with indentation") {
+                let source =
+                    "    // sourcery:inline:Type.AutoCoding\n" +
+                        "    var something: Int\n" +
+                "    // sourcery:end\n"
```
Comment (ilyapuchka) [inline SourceryTests/Parsing/TemplatesAnnotationParser+ForceParseInlineCodeSpec.swift:51]:
```suggestion
                let source = """
                        // sourcery:inline:Type.AutoCoding
                            var something: Int
                        // sourcery:end
                    """
```

File: SourceryTests/Parsing/TemplatesAnnotationParser+ForceParseInlineCodeSpec.swift [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+                    expect(annotatedRanges?.map { $0.range }).to(equal([NSRange(location: 39, length: 23)]))
+                    expect(annotatedRanges?.map { $0.indentation }).to(equal(["    "]))
+                }
+
+                it("does not remove the content between the markup when force parse parameter is set to template name") {
+                    expect(result.contents).to(equal(
+                        "    // sourcery:inline:Type.AutoCoding\n" +
+                             "    var something: Int\n" +
+                        "    // sourcery:end\n"
+                    ))
```
Comment (ilyapuchka) [inline SourceryTests/Parsing/TemplatesAnnotationParser+ForceParseInlineCodeSpec.swift:66]:
```suggestion
                    expect(result.contents).to(equal("""
                        // sourcery:inline:Type.AutoCoding
                            var something: Int
                        // sourcery:end
                    """))
```

File: README.md (hunk 4 above)
Comment (krzysztofzablocki) [reply README.md]: I think re-using same argument name for both files and inline blocks is fine, it's rare to use feature anyway

**Comment by ranvirp** [comment]
> @ranvirp can you update the changes I requested?

Done

**Comment by ranvirp** [comment]
> thanks for contributing, I think it looks good, just one small change + add changelog

already done

**Comment by krzysztofzablocki** [comment]
merged in #937

**Pull request closed without merging**